### PR TITLE
Inline During Parsing Plugin 

### PR DESCRIPTION
**Pull request by marjanasolajic**
**NativeImageInlineDuringParsingPlugin** is a plugin that analyses the graph for the resolved Java method and specifies
what should to be inlined during graph parsing before the static analysis. This plugin inlines methods which folds to a constant.

To use this plugin, add option -H:+InlineBeforeAnalysis to the native-image.

We can use the plugin for supporting newInstance() without reflection configuration, test examples: 
- [reflection-static-analysis-with-try-catch.zip](https://github.com/marjanasolajic/graal/files/5249198/reflection-static-analysis-with-try-catch.zip)
- [reflection-static-analysis-with-throws.zip](https://github.com/marjanasolajic/graal/files/5249199/reflection-static-analysis-with-throws.zip)

**Results of testing plugin on benchmark suite renaissance**

renaissance_statistics2

renaissance_statistics

### CODE REVIEW

File: substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java [hunk 1]
```diff
@@ -0,0 +1,548 @@
...
+import com.oracle.svm.hosted.phases.AnalysisGraphBuilderPhase.AnalysisBytecodeParser;
+import com.oracle.svm.hosted.phases.NativeImageInlineDuringParsingPlugin.CallSite;
+import com.oracle.svm.hosted.phases.NativeImageInlineDuringParsingPlugin.InvocationResult;
+import com.oracle.svm.hosted.phases.NativeImageInlineDuringParsingPlugin.InvocationResultInline;
+import com.oracle.svm.hosted.phases.SharedGraphBuilderPhase.SharedBytecodeParser;
+
+import jdk.vm.ci.meta.ResolvedJavaMethod;
+
+/**
+ * Plugin for specifying what is inlined during graph parsing before the static analysis. This
```
Comment (vjovanov) [inline substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java]: This comment should be way more elaborate on what is happening.

Comment (vjovanov) [reply substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java]: For example, our inlining has 250 lines of comments including ASCII pictures.

File: substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java [hunk 2]
```diff
@@ -0,0 +1,548 @@
...
+ * plugin inline methods which folds to constant and is also notified {@link #notifyBeforeInline}
+ * and {@link #notifyAfterInline} the inlining.
+ */
+public class NativeImageInlineDuringParsingPlugin implements InlineInvokePlugin {
+
+    public static class Options {
+        @Option(help = "Inline methods which folds to constant during parsing before the static analysis.") public static final HostedOptionKey<Boolean> InlineBeforeAnalysis = new HostedOptionKey<>(
```
Comment (vjovanov) [inline substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java]: put // after `Option()`

File: substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java [hunk 3]
```diff
@@ -0,0 +1,548 @@
...
+                false);
+
+    }
+
+    private final boolean analysis;
+    private final HostedProviders providers;
+
+    /* the map in which we place the result of a inlining decision */
```
Comment (vjovanov) [inline substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java]: Why? Explanation would be useful.

Comment (vjovanov) [reply substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java]: Why this shape? Also, how does this type shape handle methods with nested BCIs?

File: substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java [hunk 4]
```diff
@@ -0,0 +1,548 @@
...
+    private static final ConcurrentHashMap<AnalysisMethod, ConcurrentHashMap<Integer, InvocationResult>> dataInline = new ConcurrentHashMap<>();
+
+    public NativeImageInlineDuringParsingPlugin(boolean analysis, HostedProviders providers) {
+        this.analysis = analysis;
+        this.providers = providers;
+    }
+
+    /**
+     * Determines whether a call to a given method is to be inlined.
```
Comment (vjovanov) [inline substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java]: This comment, on the other hand, is completely useless as all of it is clear from the signature.

File: substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java [hunk 5]
```diff
@@ -0,0 +1,548 @@
...
+        if (inline instanceof InvocationResultInline) {
+            InvocationResultInline inlineData = (InvocationResultInline) inline;
+            VMError.guarantee(inlineData.callee.equals(toAnalysisMethod(method)));
+
+            VMError.guarantee(((SharedBytecodeParser) b).inlineDuringParsingState == null);
+            ((SharedBytecodeParser) b).inlineDuringParsingState = inlineData;
+
+            if (analysis) {
+                AnalysisMethod aMethod = (AnalysisMethod) method;
+                aMethod.registerAsImplementationInvoked(null);
```
Comment (vjovanov) [inline substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java]: Is it if we inlined the method?

Comment (vjovanov) [reply substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java]: Why is this needed?

File: substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java [hunk 6]
```diff
@@ -0,0 +1,548 @@
...
+                if (!aMethod.isStatic()) {
+                    ensureParsed(((AnalysisBytecodeParser) b).bb, aMethod);
```
Comment (vjovanov) [inline substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java]: Why do we need this? Again if the method is inlined, do we have to parse it?

File: substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java [hunk 7]
```diff
@@ -0,0 +1,548 @@
...
+                    if (args[0].isConstant()) {
```
Comment (vjovanov) [inline substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java]: Why is this needed?

File: substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java [hunk 8]
```diff
@@ -0,0 +1,548 @@
...
+        }
+    }
+}
+
+/**
+ * This class detects what can be inlined.
+ */
+class TrivialMethodDetector {
+
+    final BigBang bb;
```
Comment (vjovanov) [inline substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java]: All of these can be private.

File: substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java [hunk 9]
```diff
@@ -0,0 +1,548 @@
...
+        this.debug = debug;
+        this.prototypeGraphBuilderConfig = makePrototypeGraphBuilderConfig(originalGraphBuilderConfig);
+        this.options = options;
+        this.providers = providers;
+    }
+
+    private static GraphBuilderConfiguration makePrototypeGraphBuilderConfig(GraphBuilderConfiguration originalGraphBuilderConfig) {
+        GraphBuilderConfiguration result = originalGraphBuilderConfig.copy();
+
+        result.getPlugins().clearParameterPlugin();
```
Comment (vjovanov) [inline substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java]: Is this overly destructive? Are there any other parameter plugins that we might need?

File: substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java [hunk 10]
```diff
@@ -0,0 +1,548 @@
...
+
+        return result;
+    }
+
+    @SuppressWarnings("try")
+    InvocationResult analyzeMethod(CallSite callSite, AnalysisMethod method, ValueNode[] args, Object existingSingleAllowedElement) {
+        if (!method.hasBytecodes()) {
+            /* Native method. */
+            return InvocationResult.ANALYSIS_TOO_COMPLICATED;
+
```
Comment (vjovanov) [inline substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java]: Extra space.

File: substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java [hunk 11]
```diff
@@ -0,0 +1,548 @@
...
+        } else if (method.isSynchronized()) {
+            /*
+             * Synchronization operations will always bring us above the node limit, so no point in
+             * starting an analysis.
+             */
+            return InvocationResult.ANALYSIS_TOO_COMPLICATED;
+
```
Comment (vjovanov) [inline substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java]: Extra space.

File: substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java [hunk 12]
```diff
@@ -0,0 +1,548 @@
...
+        @Override
+        public String toString() {
+            return caller.format("%h.%n(%p)") + "@" + bci;
+        }
+    }
+
+    static class InvocationResult {
+        static final InvocationResult ANALYSIS_TOO_COMPLICATED = new InvocationResult();
+
```
Comment (vjovanov) [inline substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java]: Extra space.

**Comment by vjovanov** [comment]
I would clean up all the warnings in IntelliJ, except for pattern matching style we use in cascading ifs.

File: substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java [hunk 13]
```diff
@@ -0,0 +1,548 @@
...
+             */
+            return InvocationResult.ANALYSIS_TOO_COMPLICATED;
+        }
+    }
+
+    /**
+     * We collect information during graph construction using and use result of
+     * {@link #analyzeGraph} in {@link #analyzeMethod}.
+     */
+    class MethodState extends NodeEventListener implements InlineInvokePlugin {
```
Comment (vjovanov) [inline substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java]: I find it very confusing that MethodState is three things at the same time. Maybe give it a better name or simply split it up in two things that get state as input.

File: substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java [hunk 14]
```diff
@@ -0,0 +1,548 @@
...
+        @Option(help = "Inline methods which folds to constant during parsing before the static analysis.") public static final HostedOptionKey<Boolean> InlineBeforeAnalysis = new HostedOptionKey<>(
+                false);
+
+    }
+
+    private final boolean analysis;
+    private final HostedProviders providers;
+
+    /* the map in which we place the result of a inlining decision */
+    private static final ConcurrentHashMap<AnalysisMethod, ConcurrentHashMap<Integer, InvocationResult>> dataInline = new ConcurrentHashMap<>();
```
Comment (vjovanov) [inline substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java]: This is based on the assumption that every call site within a method has a unique BCI. I don't think this is a case. Have you verified this? If yes, please add an assertion that makes this assumption a fact, if no please modify the data structure to contain the whole BCI chain that leads to the call site.

File: substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java [hunk 15]
```diff
@@ -96,29 +92,27 @@
...
 
-    /* the map in which we place the result of a inlining decision */
-    private static final ConcurrentHashMap<AnalysisMethod, ConcurrentHashMap<Integer, InvocationResult>> dataInline = new ConcurrentHashMap<>();
+    /**
+     * The map in which we place the result of a inlining decision. During analysis we store
+     * information about method in separate map using {@link #callSitemap} and {@link #putIfAbsent}.
+     * After analysis we read information from this map to find inline decision we made using
+     * {@link #getResult}.
+     */
+    private static final ConcurrentHashMap<AnalysisMethod, ConcurrentHashMap<CallSite, InvocationResult>> dataInline = new ConcurrentHashMap<>();
```
Comment (marjanasolajic) [inline substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java]: https://github.com/marjanasolajic/graal/pull/13#discussion_r482010185

@vjovanov I modified the data structure to collect directly CallSite objects with the invocation result. Also, methods for adding new results and searching for the result are modified.

File: substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java (hunk 13 above)
Comment (marjanasolajic) [reply substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java]: Renamed MethodState to MethodNodeTrackingAndInline.

File: substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java (hunk 14 above)
Comment (marjanasolajic) [reply substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java]: https://github.com/marjanasolajic/graal/pull/13#discussion_r483984417

File: substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java (hunk 12 above)
Comment (marjanasolajic) [reply substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java]: removed

File: substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java (hunk 11 above)
Comment (marjanasolajic) [reply substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java]: removed

File: substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java (hunk 10 above)
Comment (marjanasolajic) [reply substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java]: removed

File: substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java (hunk 9 above)
Comment (marjanasolajic) [reply substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java]: I tested the latest version, remove or leave them, either one is fine because two of them are in the compiler package, and one is for intrinsification (we are not interfering with any intrinsic).

File: substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java (hunk 8 above)
Comment (marjanasolajic) [reply substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java]: added private keyword

File: substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java (hunk 5 above)
Comment (marjanasolajic) [reply substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java]: If I remove it, there is a parsing error for non-recursive calls in the intrinsic org.graalvm.compiler.replacements.JDK9StringSubstitutions.equals(String, Object).

File: substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java (hunk 6 above)
Comment (marjanasolajic) [reply substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java]: Yes, I overlooked that. We don't need this. This part of code and method are removed.

File: substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java (hunk 4 above)
Comment (marjanasolajic) [reply substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java]: Removed.

File: substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java (hunk 3 above)
Comment (marjanasolajic) [reply substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java]: https://github.com/marjanasolajic/graal/pull/13#discussion_r483984417

File: substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java (hunk 2 above)
Comment (marjanasolajic) [reply substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java]: corrected

File: substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java (hunk 1 above)
Comment (marjanasolajic) [reply substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java]: new comment has been added in the last commit

File: substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java [hunk 16]
```diff
@@ -0,0 +1,545 @@
+/*
+ * Copyright (c) 2013, 2017, Oracle and/or its affiliates. All rights reserved.
```
Comment (vjovanov) [inline substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java]: 2020

File: substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java (hunk 5 above)
Comment (vjovanov) [reply substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java]: What is the error? I would dig deeper on this.

File: substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java (hunk 16 above)
Comment (marjanasolajic) [reply substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java]: changed

File: substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java (hunk 5 above)
Comment (marjanasolajic) [reply substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java]: org.graalvm.compiler.debug.GraalError: All non-recursive calls in the intrinsic org.graalvm.compiler.replacements.JDK9StringSubstitutions.equals(String, Object) must be inlined or intrinsified: found call to String.coder()

Error is thrown by ReplacementsImpl for the method String.coder() if we don't call registerAsImplementationInvoked when we decide to inline this method.

Comment (marjanasolajic) [reply substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java]: We are not interfering with any intrinsic method, while ReplacementsImpl forces inlining of intrinsics.  We decide to inline simple methods like String.coder() that has only constants nodes and return node in the graph and this method is also called by String.length() that is called by intrinsic JDK9StringSubstitutions.equals().

File: substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java [hunk 17]
```diff
@@ -0,0 +1,560 @@
...
+        if (inline instanceof InvocationResultInline) {
+            InvocationResultInline inlineData = (InvocationResultInline) inline;
+            VMError.guarantee(inlineData.callee.equals(toAnalysisMethod(method)));
+
+            VMError.guarantee(((SharedBytecodeParser) b).inlineDuringParsingState == null);
+            ((SharedBytecodeParser) b).inlineDuringParsingState = inlineData;
+
+            if (analysis) {
+                AnalysisMethod aMethod = (AnalysisMethod) method;
+                aMethod.registerAsImplementationInvoked(null);
```
Comment (marjanasolajic) [inline substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java]: It's necessary to further research the reason why is this required.

File: substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java [hunk 18]
```diff
@@ -0,0 +1,609 @@
...
+            } else {
+                CallSite callSite = new CallSite(getCallingContextSubset(b, b.getDepth()), toAnalysisMethod(callee));
+                inline = inlineDuringParsingState.children.get(callSite);
+                if (inline == null) {
+                    /*
+                     * We already decided to inline the first callee into the root method, so now we
+                     * must recursively inline everything. Missing information about the child can
+                     * happen when there is a virtual invoke in the inlined callee.
+                     */
+                    inline = new InvocationResultInline(callSite, toAnalysisMethod(callee));
```
Comment (vjovanov) [inline substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java]: How can we have a virtual invoke when we allow only constants and field load?
